Modifications to `periodicity`

akash9712 opened the pull request:
`periodicity` earlier used  a check after splitting the given function
using `decompogen` to check that the function being tested for periodicity during recursion
is not same as the original function, to avoid infinite recursion. This
check equated both the functions. This, however, led to the `periodicity` being
stuck, if the decomposition of the functions were the same. Equating the
decomposition of the functions during the check solves this issue.

Example when this happened:
```
>>> from sympy import *
>>> x = Symbol('x')
>>> periodicity(gamma(S(1)/7 + 1/x))

```

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed


#### Other comments

akash9712 committed:
Modifications to `periodicity`

`periodicity` earlier used  a check after splitting the given function
using `decompogen` to check that the function being tested for periodicity
is not same as the original function, to avoid infinite recursion. This
check equated both the functions. This, however, led to the funcction being
stuck if the decomposition of the functions were the same. Equating the
decomposition of the functions during the check solves this issue.

Example when this happened:
```
>>> from sympy import *
>>> x = Symbol('x')
>>> periodicity(gamma(S(1)/7 + 1/x))

```

smichr commented on sympy/calculus/util.py:
```diff
@@ -476,7 +476,9 @@ def _check(orig_f, period):
                 g = compogen(g_s[start_index:], symbol)
-                if g != orig_f and g != f: # Fix for issue 12620
+                # Fix for issue 12620
+                if decompogen(g, symbol) != decompogen(orig_f, symbol) \
```
Would it be advantageous to factor out the orig_g calculation so that doesn't have to be repeatedly calculated for every g? Also, you are comparing `decompogen(g, symbol)` (an element-ish thing of `g_s`) with `decompogen(f, symbol)`...which *is* `g_s`. Will those two every actually be equal? And, again, is it necessary to calculate `decompogen(g, symbol)` twice to do this test?

Since g of the `for` loop is never used, please re-write that as `for index in range(len(g_s)):`...but I wonder if some edit went awry since why is the `g_s` being reversed if the `g` is never used? /!\ CAUTION

akash9712 replied:
Comparing `decompogen(g, symbol)` with `decompogen(f, symbol)`, makes sure that the function does not enter infinite recursion, because even though the decomposition of two functions might be the same, the functions might be different, as was happening in the test case added. Because the `periodicity` uses recursion to break down a function into functions for which the periodicity can be directly calculated, this causes the infinite recursion.

Removing `g` from the `for` loop and rewriting as suggested does seems less complicated. There seems to be no problem, as both the loops do the exact same thing.

If by removing the `orig_f` calculation, you mean the comparison on L479, I don't think it should be removed. `f` is getting reassigned in some places, and so comparison to `orig_f` should be done as well.

I shall correct the part where `decompogen(g, symbol)` is being computed twice.

akash9712 committed:
Changes to `periodicity`

Removed unnecessary reuse of `decompogen` function, and simplified the `for`
loop in which it was being used.

akash9712 commented:
@ping smichr

akash9712 commented:
ping @smichr

czgdp1807 commented:
```python
Python 3.6.8 (default, Aug 20 2019, 17:12:48) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> x = Symbol('x')
>>> periodicity(gamma(S(1)/7 + 1/x), x)
>>> z = periodicity(gamma(S(1)/7 + 1/x), x)
>>> print(z)
None
>>> 
```
Latest update is above code. @smichr Is this PR still needed?

czgdp1807 commented:
Closing this for now, in reference to, https://github.com/sympy/sympy/pull/13896#issuecomment-531537603 
Thanks for the contributions. Please let us know if this can be re-opened again.
